grpc-compiler-plugin/src/main/java/org/ballerinalang/net/grpc/plugin/GrpcServiceValidator.java

```diff
@@ -0,0 +1,220 @@
+import io.ballerina.compiler.syntax.tree.Node;
+import io.ballerina.compiler.syntax.tree.NodeList;
+import io.ballerina.compiler.syntax.tree.ParameterNode;
+import io.ballerina.compiler.syntax.tree.RequiredParameterNode;
+import io.ballerina.compiler.syntax.tree.RestParameterNode;
+import io.ballerina.compiler.syntax.tree.SeparatedNodeList;
+import io.ballerina.compiler.syntax.tree.ServiceDeclarationNode;
+import io.ballerina.compiler.syntax.tree.SyntaxKind;
+import io.ballerina.projects.plugins.AnalysisTask;
+import io.ballerina.projects.plugins.SyntaxNodeAnalysisContext;
+import io.ballerina.tools.diagnostics.Diagnostic;
+import io.ballerina.tools.diagnostics.DiagnosticFactory;
+import io.ballerina.tools.diagnostics.DiagnosticInfo;
+import io.ballerina.tools.diagnostics.DiagnosticSeverity;
+import org.ballerinalang.net.grpc.MessageUtils;
+
+import java.util.List;
+import java.util.Locale;
+import java.util.Optional;
+
+/**
+ * gRPC service validator for compiler API.
+ */
+public class GrpcServiceValidator implements AnalysisTask<SyntaxNodeAnalysisContext> {
+
+    private static final String GRPC_GENERIC_CALLER = "caller";
+
+    @Override
+    public void perform(SyntaxNodeAnalysisContext syntaxNodeAnalysisContext) {
+
+        ServiceDeclarationNode serviceDeclarationNode = (ServiceDeclarationNode) syntaxNodeAnalysisContext.node();
+        Optional<Symbol> optionalServiceDeclarationSymbol = syntaxNodeAnalysisContext.semanticModel()
+                .symbol(serviceDeclarationNode);
+        if (optionalServiceDeclarationSymbol.isPresent() &&
+                (optionalServiceDeclarationSymbol.get() instanceof ServiceDeclarationSymbol)) {
+
+            ServiceDeclarationSymbol serviceDeclarationSymbol = (ServiceDeclarationSymbol)
+                    optionalServiceDeclarationSymbol.get();
+
+            if (isBallerinaGrpcService(serviceDeclarationSymbol)) {
+                String serviceName = serviceNameFromServiceDeclarationNode(serviceDeclarationNode);
+                validateServiceAnnotation(serviceDeclarationNode, syntaxNodeAnalysisContext, serviceDeclarationSymbol);
+                serviceDeclarationNode.members().stream().filter(child ->
+                        child.kind() == SyntaxKind.OBJECT_METHOD_DEFINITION
+                                || child.kind() == SyntaxKind.RESOURCE_ACCESSOR_DEFINITION).forEach(node -> {
+
+                    FunctionDefinitionNode functionDefinitionNode = (FunctionDefinitionNode) node;
+                    // Check functions are remote or not
+                    validateServiceFunctions(functionDefinitionNode, syntaxNodeAnalysisContext);
+                    // Check params and return types
+                    validateFunctionSignature(functionDefinitionNode, syntaxNodeAnalysisContext, serviceName);
+
+                });
+            }
+
+        }
+
+    }
+
+    public boolean isBallerinaGrpcService(ServiceDeclarationSymbol serviceDeclarationSymbol) {
```

Comment:
Shall we make these local functions `private`

Reply:
Fixed in here https://github.com/ballerina-platform/module-ballerina-grpc/pull/154/commits/b419385aa5fdaaa4d2bfafdf82ef1093e5f0b9c3